CPS-0010/README.md

```diff
@@ -0,0 +1,276 @@
+
+This design ensures that wallets can function without a query layer and thus without runtime infrastructure needed.
+
+##### Own-data wallets
+
+Own-data wallets provide chain queries related to users' own data and wallet state (users' addresses and utxos).
+CIP-30 falls into this category.
+
+![diagram showing possible wallet and dApp architecture for own-data wallets](./own-data-wallet.drawio.png)
+
+This architecture has two major issues:
+
+- **Unclear separation of concerns** - while wallets provide certain functions for dApp developers, they have no means to enforce their use, and naturally some developers opt to using their own query layers when they see fit, with consequences that are hard to predict. For example, the wallet may not allow signing a transaction that consumes an UTxO it considers "locked".
+
+- **Two sources of truth problem** - on Cardano, every running node has its own opinion on the set of currently unspent transaction outputs. Only [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency) is guaranteed. Any dApp that interacts with an own-data wallet has to deal with the inconsistency between local `cardano-node`-based query layer and light wallet query layer, especially when dApp workflow involves sending multiple interactions with the wallet in quick succession. For example, a wallet may refuse to sign a transaction containing an UTxO it does not (yet) know about.
+
+Even if we enforce clear separation of concerns in the standards, e.g. by stating that sending a transaction bypassing the wallet is not allowed, the problem of two sources of truth will remain.
+
+##### Full-data wallets
+
+Full-data wallets allow to query blockchain data outside of user's scope (i.e. anything not covered by own-data wallets).
+
+![diagram showing possible wallet and dApp architecture for full-data wallets](./full-data-wallet.drawio.png)
+
+Depending on dApp needs, full-data wallets open a way to implement fully-functional dApps that use non-local blockchain data without the need for the developer to maintain dApp backend infrastructure. As a result, more decentralized and censorship-resistant architectures become possible: a dApp can be distributed as a set of files, and deployed to IPFS or static hosting websites.
+
+Full-data wallets enable "single source of truth" architecture: no need to work around data inconsistency between two query layers.
+
+They are more expensive to operate due to higher query layer requirements, but the risks could be lowered by letting query layer providers and wallet developers be separate entities. That would also allow the end users to choose query layer providers they trust more rather than being forced to trust wallet's infrastructure as well as code, and, optionally, use their own query layer deployments, thus alleviating the need to trust any external wallet backend while enjoying the usual interface.
```

Comment:
```suggestion
They are more expensive to operate due to higher query layer requirements, but the risks could be lowered by letting query layer providers and wallet developers be separate entities. That would also allow the end users to choose query layer providers they trust more, rather than being forced to trust wallet's infrastructure as well as code, and, optionally, use their own query layer deployments, thus alleviating the need to trust any external wallet backend while enjoying the usual interface.
```
(grammar, and also resolves an ambiguity in the reading) matches the comma at the end of this phrase.